### PR TITLE
feat(raredisease) add sex check

### PR DESCRIPTION
**Pull request by peterpru**
## Description
This PR adds a sex check to `cg workflow raredisease metrics-deliver case`, comparing the sex in statusDB against the predicted sex in peddy.

### Added

- Sex check as part of cg workflow raredisease metrics-deliver case

### Changed

-

### Fixed

-

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b add-raredisease-sex-relatedness-check
    ```

### How to test

- [x] Do `cg workflow raredisease metrics-deliver funkitten`, it should complete successfully
- [x] Change predicted sex from male to female in two samples: `sed -i 's/"predicted_sex_sex_check": "male",/"predicted_sex_sex_check": "female",/g' multiqc/multiqc_data/multiqc_data.json`
- [x] Do `cg workflow raredisease metrics-deliver funkitten`, it should fail.

### Expected test outcome

- [x] Check that code completes as normal:
image

- [x] Check that code fails as two males changed to females:
image

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/workflow/raredisease.py [hunk 1]
```diff
@@ -111,5 +111,7 @@ def get_managed_variants(self) -> list[str]:
         """Create and return the managed variants."""
         return self._get_managed_variants(genome_build=GenePanelGenomeBuild.hg19)
 
-    def get_workflow_metrics(self) -> dict:
+    def get_workflow_metrics(self, sample_id: str) -> dict:
+        sample: Sample = self.status_db.get_sample_by_internal_id(internal_id=sample_id)
+        RAREDISEASE_METRIC_CONDITIONS["predicted_sex_sex_check"]["threshold"] = sample.sex
         return RAREDISEASE_METRIC_CONDITIONS
```
Comment (ivadym) [inline cg/meta/workflow/raredisease.py:117]: We don't want to modify the original constant. Create a copy of it instead (either with `dict()` or `.copy()`):

```suggestion
        sample: Sample = self.status_db.get_sample_by_internal_id(internal_id=sample_id)
        metric_conditions: dict[str, dict[str, Any]] = dict(RAREDISEASE_METRIC_CONDITIONS)
        metric_conditions["predicted_sex_sex_check"]["threshold"] = sample.sex
        return metric_conditions
```

**Review by ivadym — APPROVED** [review]
Smart solution 💯

File: cg/constants/nf_analysis.py [hunk 2]
```diff
@@ -21,6 +21,7 @@ class NfTowerStatus(StrEnum):
     "MEDIAN_TARGET_COVERAGE": {"norm": "gt", "threshold": 26},
     "PCT_TARGET_BASES_10X": {"norm": "gt", "threshold": 0.95},
     "PCT_EXC_ADAPTER": {"norm": "lt", "threshold": 0.0005},
+    "predicted_sex_sex_check": {"norm": "eq", "threshold": None},
```
Comment (henrikstranneheim) [inline cg/constants/nf_analysis.py:24]:
```suggestion
    "sex_check_predicted_sex": {"norm": "eq", "threshold": None},
```

File: cg/meta/workflow/raredisease.py [hunk 3]
```diff
@@ -111,5 +111,8 @@ def get_managed_variants(self) -> list[str]:
         """Create and return the managed variants."""
         return self._get_managed_variants(genome_build=GenePanelGenomeBuild.hg19)
 
-    def get_workflow_metrics(self) -> dict:
-        return RAREDISEASE_METRIC_CONDITIONS
+    def get_workflow_metrics(self, sample_id: str) -> dict:
+        sample: Sample = self.status_db.get_sample_by_internal_id(internal_id=sample_id)
+        metric_conditions: dict[str, dict[str, Any]] = dict(RAREDISEASE_METRIC_CONDITIONS)
+        metric_conditions["predicted_sex_sex_check"]["threshold"] = sample.sex
```
Comment (henrikstranneheim) [inline cg/meta/workflow/raredisease.py:117]: Can we make this a separate function to encapsulate the logic better?

**Review by henrikstranneheim — COMMENTED** [review]
Nice, 👍 .

File: cg/constants/nf_analysis.py (hunk 2 above)
Comment (peterpru) [reply cg/constants/nf_analysis.py:24]: I agree with you, but this is what the variable is called in the multiqc json output file, so I will have to keep it as is.

Comment (henrikstranneheim) [reply cg/constants/nf_analysis.py:24]: I was afraid of that, 👍 .

File: cg/meta/workflow/raredisease.py [hunk 4]
```diff
@@ -111,5 +111,11 @@ def get_managed_variants(self) -> list[str]:
...
 
-    def get_workflow_metrics(self) -> dict:
-        return RAREDISEASE_METRIC_CONDITIONS
+    def get_workflow_metrics(self, sample_id: str) -> dict:
+        sample: Sample = self.status_db.get_sample_by_internal_id(internal_id=sample_id)
+        metric_conditions: dict[str, dict[str, Any]] = dict(RAREDISEASE_METRIC_CONDITIONS)
+        self.perform_sex_check(sample, metric_conditions)
+        return metric_conditions
+
+    def perform_sex_check(self, sample: Sample, metric_conditions: dict) -> None:
```
Comment (ivadym) [inline cg/meta/workflow/raredisease.py:120]: rename the method to something that better reflects what it's doing, for example:

```suggestion
    def set_order_sex_for_sample(self, sample: Sample, metric_conditions: dict) -> None:
```

Comment (ivadym) [reply cg/meta/workflow/raredisease.py:120]: Maybe even better: `update_metric_conditions_with_sample_sex` or `update_metric_conditions_with_order_sex`

File: cg/meta/workflow/raredisease.py [hunk 5]
```diff
@@ -111,5 +111,11 @@ def get_managed_variants(self) -> list[str]:
         """Create and return the managed variants."""
         return self._get_managed_variants(genome_build=GenePanelGenomeBuild.hg19)
 
-    def get_workflow_metrics(self) -> dict:
-        return RAREDISEASE_METRIC_CONDITIONS
+    def get_workflow_metrics(self, sample_id: str) -> dict:
+        sample: Sample = self.status_db.get_sample_by_internal_id(internal_id=sample_id)
+        metric_conditions: dict[str, dict[str, Any]] = dict(RAREDISEASE_METRIC_CONDITIONS)
```
Comment (ivadym) [inline cg/meta/workflow/raredisease.py:116]: Any is not imported

**Comment by peterpru** [comment]
Successful deployment in stage:
image

Successful deployment in production:
image

**Pull request merged**